<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ delimtools</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://legallab.github.io/delimtools/logo.svg"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">delimtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/install.html">Installing Species Delimitation Softwares</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/legalLab/delimtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="abgd_tbl.html">abgd_tbl()</a></code>

        </dt>
        <dd>A Command-Line Interface for ABGD - Automatic Barcode Gap Discovery</dd>
      </dl><dl><dt>

          <code><a href="as_dwc.html">as_dwc()</a></code>

        </dt>
        <dd>Rename Columns using Darwin Core Standard Terms</dd>
      </dl><dl><dt>

          <code><a href="asap_tbl.html">asap_tbl()</a></code>

        </dt>
        <dd>A Command-Line Interface for ASAP - Assemble Species by Automatic Partitioning</dd>
      </dl><dl><dt>

          <code><a href="bgmyc_tbl.html">bgmyc_tbl()</a></code>

        </dt>
        <dd>Turns bGMYC Results Into a Tibble</dd>
      </dl><dl><dt>

          <code><a href="check_delim.html">check_delim()</a></code>

        </dt>
        <dd>Checks If Two or More Species Delimitation Outputs are (Nearly) Equal</dd>
      </dl><dl><dt>

          <code><a href="check_identifiers.html">check_identifiers()</a></code>

        </dt>
        <dd>Checks for Differences Between Identifiers in Metadata and DNA Sequence Files</dd>
      </dl><dl><dt>

          <code><a href="clean_dna.html">clean_dna()</a></code>

        </dt>
        <dd>Removes Gaps, Ambiguities and Missing Data from DNA Sequences</dd>
      </dl><dl><dt>

          <code><a href="collapse_others.html">collapse_others()</a></code>

        </dt>
        <dd>Summarise Haplotype Metadata Down to One Row</dd>
      </dl><dl><dt>

          <code><a href="confidence_intervals.html">gmyc_ci()</a></code> <code><a href="confidence_intervals.html">bgmyc_ci()</a></code> <code><a href="confidence_intervals.html">locmin_ci()</a></code> <code><a href="confidence_intervals.html">mptp_ci()</a></code>

        </dt>
        <dd>Confidence Intervals for Species Delimitations Methods</dd>
      </dl><dl><dt>

          <code><a href="delim_autoplot.html">delim_autoplot()</a></code>

        </dt>
        <dd>Plot Phylogenetic Trees With Species Delimitation Partitions</dd>
      </dl><dl><dt>

          <code><a href="delim_autoplot2.html">delim_autoplot2()</a></code>

        </dt>
        <dd>Plot Phylogenetic Trees With Species Delimitation Partitions</dd>
      </dl><dl><dt>

          <code><a href="delim_brewer.html">delim_brewer()</a></code>

        </dt>
        <dd>Customize Delimitation Colors</dd>
      </dl><dl><dt>

          <code><a href="delim_consensus.html">delim_consensus()</a></code>

        </dt>
        <dd>Estimate a Majority-Vote Consensus</dd>
      </dl><dl><dt>

          <code><a href="delim_join.html">delim_join()</a></code>

        </dt>
        <dd>Join Multiple Species Delimitation Methods Outputs</dd>
      </dl><dl><dt>

          <code><a href="delimtools-package.html">delimtools</a></code> <code><a href="delimtools-package.html">delimtools-package</a></code>

        </dt>
        <dd>Helper Functions for Species Delimitation Analysis</dd>
      </dl><dl><dt>

          <code><a href="drop_sequences.html">drop_sequences()</a></code>

        </dt>
        <dd>Remove Sequences of a DNAbin list object</dd>
      </dl><dl><dt>

          <code><a href="dwc_terms.html">dwc_terms()</a></code>

        </dt>
        <dd>Print Darwin Core Terms, Definitions and Examples as Bullet Lists</dd>
      </dl><dl><dt>

          <code><a href="geophagus.html">geophagus</a></code>

        </dt>
        <dd>Cytochrome C Oxidase Sequences of Geophagus Eartheaters</dd>
      </dl><dl><dt>

          <code><a href="geophagus_beast.html">geophagus_beast</a></code>

        </dt>
        <dd>Geophagus Eartheaters Ultrametric Tree</dd>
      </dl><dl><dt>

          <code><a href="geophagus_bootstraps.html">geophagus_bootstraps</a></code>

        </dt>
        <dd>Geophagus Eartheaters Bootstrap Trees</dd>
      </dl><dl><dt>

          <code><a href="geophagus_delims.html">geophagus_delims</a></code>

        </dt>
        <dd>Geophagus Eartheaters Species Partitions</dd>
      </dl><dl><dt>

          <code><a href="geophagus_info.html">geophagus_info</a></code>

        </dt>
        <dd>Geophagus Earthearts Associated Metadata</dd>
      </dl><dl><dt>

          <code><a href="geophagus_posterior.html">geophagus_posterior</a></code>

        </dt>
        <dd>Geophagus Eartheaters Posterior Trees</dd>
      </dl><dl><dt>

          <code><a href="geophagus_raxml.html">geophagus_raxml</a></code>

        </dt>
        <dd>Geophagus Eartheaters Phylogram</dd>
      </dl><dl><dt>

          <code><a href="get_delim_cols.html">get_delim_cols()</a></code>

        </dt>
        <dd>Extract Labels and Colors from Species Delimitation Partitions</dd>
      </dl><dl><dt>

          <code><a href="get_dwc.html">get_dwc()</a></code>

        </dt>
        <dd>Get Darwin Core Terms and Definitions</dd>
      </dl><dl><dt>

          <code><a href="gmyc_tbl.html">gmyc_tbl()</a></code>

        </dt>
        <dd>Turns GMYC Results Into a Tibble</dd>
      </dl><dl><dt>

          <code><a href="hap_collapse.html">hap_collapse()</a></code>

        </dt>
        <dd>Removes Duplicated Sequences from Alignment</dd>
      </dl><dl><dt>

          <code><a href="hap_unite.html">hap_unite()</a></code>

        </dt>
        <dd>Unite Haplotype Summaries with Species Delimitation Outputs</dd>
      </dl><dl><dt>

          <code><a href="haplotype_tbl.html">haplotype_tbl()</a></code>

        </dt>
        <dd>Summarise Haplotypes Down to One Row</dd>
      </dl><dl><dt>

          <code><a href="locmin_tbl.html">locmin_tbl()</a></code>

        </dt>
        <dd>Turns Local Minima Results into a Tibble</dd>
      </dl><dl><dt>

          <code><a href="match_ratio.html">match_ratio()</a></code>

        </dt>
        <dd>Compute Agreement Between Alternative Species Delimitation Partitions</dd>
      </dl><dl><dt>

          <code><a href="min_brlen.html">min_brlen()</a></code>

        </dt>
        <dd>A function to report the smallest tip-to-tip distances in a phylogenetic tree</dd>
      </dl><dl><dt>

          <code><a href="morph_tbl.html">morph_tbl()</a></code>

        </dt>
        <dd>Generating a Morphological Delimitation Table</dd>
      </dl><dl><dt>

          <code><a href="mptp_tbl.html">mptp_tbl()</a></code>

        </dt>
        <dd>A Command-Line Interface for mPTP - multi-rate Poisson Tree Processes</dd>
      </dl><dl><dt>

          <code><a href="report_delim.html">report_delim()</a></code>

        </dt>
        <dd>Report Unique Species Partitions</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pedro Bittencourt, Rupert Collins, Tomas Hrbek.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

