<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Introduction to delimtools • delimtools</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to delimtools">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">delimtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/delimtools.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/install.html">Installing Species Delimitation Softwares</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/legalLab/delimtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.svg" class="logo" alt=""><h1>Introduction to delimtools</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/legalLab/delimtools/blob/HEAD/vignettes/articles/delimtools.Rmd" class="external-link"><code>vignettes/articles/delimtools.Rmd</code></a></small>
      <div class="d-none name"><code>delimtools.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="section"></h3>
</div>
<div class="section level2">
<h2 id="rationale">Rationale<a class="anchor" aria-label="anchor" href="#rationale"></a>
</h2>
<p>There are at least three major steps when doing single-locus species
delimitation analysis:</p>
<ul>
<li><p>Preparing your data;</p></li>
<li><p>Running species delimitation analysis;</p></li>
<li><p>Visualizing results.</p></li>
</ul>
<p>The rationale of <code>delimtools</code> is to provide helper
functions for all these steps. This guide introduces you to these
functions and shows how to apply them to your datasets. In this
tutorial, we will use the <code>geophagus</code> dataset provided in
this package. This dataset contains 354 cytochrome c oxidase sequences
of the <em>Geophagus sensu stricto</em> species group downloaded from
GenBank. Most of these sequences are from the data analysed by Ximenes
et al. (2021) and is documented in <code><a href="../reference/geophagus.html">?geophagus</a></code>.</p>
<div class="section level3">
<h3 id="preparing-your-data">Preparing your data<a class="anchor" aria-label="anchor" href="#preparing-your-data"></a>
</h3>
<p>Usually, our main source of information for single-locus species
delimitation analysis is a <a href="https://www.ncbi.nlm.nih.gov/genbank/fastaformat/" class="external-link">FASTA</a>
formated file containing nucleotide sequences. The most straighforward
way to import a FASTA formated file to R Environment is by using
<code><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">ape::read.FASTA()</a></code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path_to_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/geophagus.fasta"</span>, package <span class="op">=</span> <span class="st">"delimtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">geophagus</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.dna.html" class="external-link">read.FASTA</a></span><span class="op">(</span>file<span class="op">=</span> <span class="va">path_to_file</span><span class="op">)</span></span></code></pre></div>
<p>Since <code>geophagus</code> dataset is provided along with
<code>delimtools</code>, one can simply call it directly:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geophagus</span></span>
<span><span class="co">#&gt; 354 DNA sequences in binary format stored in a list.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; All sequences of same length: 690 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labels:</span></span>
<span><span class="co">#&gt; MZ504301.1</span></span>
<span><span class="co">#&gt; MZ504328.1</span></span>
<span><span class="co">#&gt; MZ504318.1</span></span>
<span><span class="co">#&gt; MZ504313.1</span></span>
<span><span class="co">#&gt; MZ504311.1</span></span>
<span><span class="co">#&gt; MZ504312.1</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Base composition:</span></span>
<span><span class="co">#&gt;     a     c     g     t </span></span>
<span><span class="co">#&gt; 0.237 0.282 0.178 0.303 </span></span>
<span><span class="co">#&gt; (Total: 244.26 kb)</span></span></code></pre></div>
<div class="section level4">
<h4 id="check-identifiers-across-files">Check identifiers across files<a class="anchor" aria-label="anchor" href="#check-identifiers-across-files"></a>
</h4>
<p>Although the header of each sequence in a FASTA file may contain
information about the source organism from which the sequence was
obtained, we recommend storing this information in a tabular format file
(e.g. <code>csv</code>, <code>tsv</code>, <code>xlsx</code>, etc.),
using an unique sequence identifier for both FASTA and tabular files. We
can have a glimpse of our sequence metadata by checking
<code>geophagus_info</code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">geophagus_info</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 354</span></span>
<span><span class="co">#&gt; Columns: 19</span></span>
<span><span class="co">#&gt; $ scientificName        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Geophagus_altifrons"<span style="color: #949494;">, </span>"Geophagus_altifrons"<span style="color: #949494;">, </span>"G…</span></span>
<span><span class="co">#&gt; $ scientificNameGenBank <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Geophagus altifrons"<span style="color: #949494;">, </span>"Geophagus altifrons"<span style="color: #949494;">, </span>"G…</span></span>
<span><span class="co">#&gt; $ class                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Teleostei"<span style="color: #949494;">, </span>"Teleostei"<span style="color: #949494;">, </span>"Teleostei"<span style="color: #949494;">, </span>"Teleoste…</span></span>
<span><span class="co">#&gt; $ order                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Cichliformes"<span style="color: #949494;">, </span>"Cichliformes"<span style="color: #949494;">, </span>"Cichliformes"<span style="color: #949494;">, </span>…</span></span>
<span><span class="co">#&gt; $ family                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Cichlidae"<span style="color: #949494;">, </span>"Cichlidae"<span style="color: #949494;">, </span>"Cichlidae"<span style="color: #949494;">, </span>"Cichlida…</span></span>
<span><span class="co">#&gt; $ genus                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Geophagus"<span style="color: #949494;">, </span>"Geophagus"<span style="color: #949494;">, </span>"Geophagus"<span style="color: #949494;">, </span>"Geophagu…</span></span>
<span><span class="co">#&gt; $ dbid                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 2063684375<span style="color: #949494;">, </span>2063684465<span style="color: #949494;">, </span>2063684523<span style="color: #949494;">, </span>2063684453<span style="color: #949494;">, </span>…</span></span>
<span><span class="co">#&gt; $ gbAccession           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "MZ504486.1"<span style="color: #949494;">, </span>"MZ504531.1"<span style="color: #949494;">, </span>"MZ504560.1"<span style="color: #949494;">, </span>"MZ504…</span></span>
<span><span class="co">#&gt; $ gene                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "coi"<span style="color: #949494;">, </span>"coi"<span style="color: #949494;">, </span>"coi"<span style="color: #949494;">, </span>"coi"<span style="color: #949494;">, </span>"coi"<span style="color: #949494;">, </span>"coi"<span style="color: #949494;">, </span>"coi"<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ length                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 659<span style="color: #949494;">, </span>699<span style="color: #949494;">, </span>627<span style="color: #949494;">, </span>612<span style="color: #949494;">, </span>552<span style="color: #949494;">, </span>654<span style="color: #949494;">, </span>627<span style="color: #949494;">, </span>651<span style="color: #949494;">, </span>666<span style="color: #949494;">, </span>609…</span></span>
<span><span class="co">#&gt; $ organelle             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "mitochondrion"<span style="color: #949494;">, </span>"mitochondrion"<span style="color: #949494;">, </span>"mitochondrion…</span></span>
<span><span class="co">#&gt; $ catalogNumber         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "1614"<span style="color: #949494;">, </span>"17578"<span style="color: #949494;">, </span>"1616"<span style="color: #949494;">, </span>"1194"<span style="color: #949494;">, </span>"2748"<span style="color: #949494;">, </span>"1610"<span style="color: #949494;">,</span>…</span></span>
<span><span class="co">#&gt; $ country               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Brazil: Tapajós- Jacareacanga- Pará"<span style="color: #949494;">, </span>"Brazil: …</span></span>
<span><span class="co">#&gt; $ publishedAs           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Mapping the hidden diversity of the Geophagus s…</span></span>
<span><span class="co">#&gt; $ publishedIn           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Unpublished"<span style="color: #949494;">, </span>"Unpublished"<span style="color: #949494;">, </span>"Unpublished"<span style="color: #949494;">, </span>"Un…</span></span>
<span><span class="co">#&gt; $ publishedBy           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "Ximenes,A.M."<span style="color: #949494;">, </span>"Ximenes,A.M."<span style="color: #949494;">, </span>"Ximenes,A.M."<span style="color: #949494;">, </span>…</span></span>
<span><span class="co">#&gt; $ date                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "25-OCT-2021"<span style="color: #949494;">, </span>"25-OCT-2021"<span style="color: #949494;">, </span>"25-OCT-2021"<span style="color: #949494;">, </span>"25…</span></span>
<span><span class="co">#&gt; $ decimalLatitude       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -6.276008<span style="color: #949494;">, </span>-3.449100<span style="color: #949494;">, </span>-6.276008<span style="color: #949494;">, </span>-1.501489<span style="color: #949494;">, </span>-6.1…</span></span>
<span><span class="co">#&gt; $ decimalLongitude      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> -57.74000<span style="color: #949494;">, </span>-57.73533<span style="color: #949494;">, </span>-57.74000<span style="color: #949494;">, </span>-52.68211<span style="color: #949494;">, </span>-48.…</span></span></code></pre></div>
<p>It’s desirable that both our FASTA file and our metadata have the
same dimensions and no abscent, mistyped, or duplicated sequence
identifiers within and between these files. Let’s check if our FASTA
file and our metadata has the same sequence identifiers using
<code><a href="../reference/check_identifiers.html">delimtools::check_identifiers()</a></code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/check_identifiers.html">check_identifiers</a></span><span class="op">(</span>data<span class="op">=</span> <span class="va">geophagus_info</span>, identifier<span class="op">=</span> <span class="st">"gbAccession"</span>, dna<span class="op">=</span> <span class="va">geophagus</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Identifiers are the same across files.</span></span></code></pre></div>
<p>Now, let’s omit the first 10 sequences and check again:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/check_identifiers.html">check_identifiers</a></span><span class="op">(</span><span class="va">geophagus_info</span>, <span class="st">"gbAccession"</span>, <span class="va">geophagus</span><span class="op">[</span><span class="fl">11</span><span class="op">:</span><span class="fl">354</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00; font-weight: bold;">Error</span><span style="font-weight: bold;"> in `check_identifiers()`:</span></span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Identifiers are not of equal length:</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> You've supplied inputs with size lengths of 354 and 344.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Identifiers absent in `geophagus_info`:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Identifiers absent in `geophagus[11:354]`:</span></span>
<span><span class="co">#&gt; MZ504313.1, MZ504328.1, MZ504311.1, MZ504299.1, MZ504318.1, MZ504309.1,</span></span>
<span><span class="co">#&gt; MZ504301.1, MZ504337.1, MZ504312.1, MZ504341.1</span></span></code></pre></div>
<p><strong>Notice</strong> that <code>check_identifiers</code> will not
evaluate if sequence identifiers in FASTA and metadata appears in the
same order:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/identical.html" class="external-link">identical</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">geophagus</span><span class="op">)</span>, <span class="va">geophagus_info</span><span class="op">$</span><span class="va">gbAccession</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.equal.html" class="external-link">all.equal</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">geophagus</span><span class="op">)</span>, <span class="va">geophagus_info</span><span class="op">$</span><span class="va">gbAccession</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "351 string mismatches"</span></span></code></pre></div>
<p>Many other functions of <code>delimtools</code> relies on a tabular
format file like <code>geophagus_info</code>, so please make sure to
have one prior to any analysis.</p>
</div>
<div class="section level4">
<h4 id="cleaning-dna-and-haplotype-collapsing">Cleaning DNA and Haplotype Collapsing<a class="anchor" aria-label="anchor" href="#cleaning-dna-and-haplotype-collapsing"></a>
</h4>
<p>Sometimes, our alignment file may contain non <em>ACTG </em> bases,
like ambiguities (N, R, Y, etc), gaps (-), or missing data (?) which may
have unintended consequences later. Using <code>clean_dna</code>, we can
find which sequences may need further manual checking.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/clean_dna.html">clean_dna</a></span><span class="op">(</span><span class="va">geophagus</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: <span style="color: #BBBB00;">⚠</span> You have missing data <span style="color: #0000BB;">"('N','-' '?')"</span> or ambiguity inside your sequence, i.e.</span></span>
<span><span class="co">#&gt; not padding the ends, and this may have unintended consequences later, as they</span></span>
<span><span class="co">#&gt; have now been removed!</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The names of the samples are bellow.</span></span>
<span><span class="co">#&gt; GU701784.1, GU701785.1</span></span>
<span><span class="co">#&gt; 354 DNA sequences in binary format stored in a list.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mean sequence length: 646.198 </span></span>
<span><span class="co">#&gt;    Shortest sequence: 505 </span></span>
<span><span class="co">#&gt;     Longest sequence: 690 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labels:</span></span>
<span><span class="co">#&gt; MZ504301.1</span></span>
<span><span class="co">#&gt; MZ504328.1</span></span>
<span><span class="co">#&gt; MZ504318.1</span></span>
<span><span class="co">#&gt; MZ504313.1</span></span>
<span><span class="co">#&gt; MZ504311.1</span></span>
<span><span class="co">#&gt; MZ504312.1</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Base composition:</span></span>
<span><span class="co">#&gt;     a     c     g     t </span></span>
<span><span class="co">#&gt; 0.237 0.282 0.178 0.303 </span></span>
<span><span class="co">#&gt; (Total: 228.75 kb)</span></span></code></pre></div>
<p><strong>Notice</strong> that <code>clean_dna</code> will remove both
leading and trailing gaps from the alignment, being necessary to realign
the file after that.</p>
<p>The presence of identical sequences (haplotypes) in our dataset may
overestimate the number of lineages detected by tree-based species
delimitation methods (e.g. bGMYC, GMYC, PTP, mPTP). Prior to
phylogenetic tree inference for these analyses, it is a best practice to
remove (collapse) these sequences into its unique haplotypes. To
collapse an alignment into its unique haplotypes, you may use the
<code>hap_collapse</code> function with default settings:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hap_collapse.html">hap_collapse</a></span><span class="op">(</span><span class="va">geophagus</span>, clean<span class="op">=</span> <span class="cn">TRUE</span>, collapseSubstrings <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: <span style="color: #BBBB00;">⚠</span> You have missing data <span style="color: #0000BB;">"('N','-' '?')"</span> or ambiguity inside your sequence, i.e.</span></span>
<span><span class="co">#&gt; not padding the ends, and this may have unintended consequences later, as they</span></span>
<span><span class="co">#&gt; have now been removed!</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The names of the samples are bellow.</span></span>
<span><span class="co">#&gt; GU701784.1, GU701785.1</span></span>
<span><span class="co">#&gt; 137 DNA sequences in binary format stored in a list.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mean sequence length: 643.007 </span></span>
<span><span class="co">#&gt;    Shortest sequence: 505 </span></span>
<span><span class="co">#&gt;     Longest sequence: 690 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labels:</span></span>
<span><span class="co">#&gt; MZ504301.1</span></span>
<span><span class="co">#&gt; MZ504318.1</span></span>
<span><span class="co">#&gt; MZ504341.1</span></span>
<span><span class="co">#&gt; MZ504337.1</span></span>
<span><span class="co">#&gt; MZ504342.1</span></span>
<span><span class="co">#&gt; MZ504304.1</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Base composition:</span></span>
<span><span class="co">#&gt;     a     c     g     t </span></span>
<span><span class="co">#&gt; 0.238 0.281 0.177 0.304 </span></span>
<span><span class="co">#&gt; (Total: 88.09 kb)</span></span></code></pre></div>
<p><strong>Notice</strong> that from the initial 354 DNA sequences,
<code>hap_collapse</code> returned a total of 137 unique DNA sequences.
This result may vary if setting <code>collapseSubstrings</code> to
<code>FALSE</code>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/hap_collapse.html">hap_collapse</a></span><span class="op">(</span><span class="va">geophagus</span>, clean<span class="op">=</span><span class="cn">TRUE</span>, collapseSubstrings <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: <span style="color: #BBBB00;">⚠</span> You have missing data <span style="color: #0000BB;">"('N','-' '?')"</span> or ambiguity inside your sequence, i.e.</span></span>
<span><span class="co">#&gt; not padding the ends, and this may have unintended consequences later, as they</span></span>
<span><span class="co">#&gt; have now been removed!</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The names of the samples are bellow.</span></span>
<span><span class="co">#&gt; GU701784.1, GU701785.1</span></span>
<span><span class="co">#&gt; 246 DNA sequences in binary format stored in a list.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Mean sequence length: 639.764 </span></span>
<span><span class="co">#&gt;    Shortest sequence: 505 </span></span>
<span><span class="co">#&gt;     Longest sequence: 690 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Labels:</span></span>
<span><span class="co">#&gt; MZ504301.1</span></span>
<span><span class="co">#&gt; MZ504328.1</span></span>
<span><span class="co">#&gt; MZ504318.1</span></span>
<span><span class="co">#&gt; MZ504341.1</span></span>
<span><span class="co">#&gt; MZ504337.1</span></span>
<span><span class="co">#&gt; MZ504299.1</span></span>
<span><span class="co">#&gt; ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Base composition:</span></span>
<span><span class="co">#&gt;     a     c     g     t </span></span>
<span><span class="co">#&gt; 0.237 0.282 0.177 0.303 </span></span>
<span><span class="co">#&gt; (Total: 157.38 kb)</span></span></code></pre></div>
<p>After setting <code>collapseSubstrings</code> to <code>FALSE</code>,
we obtained 246 unique DNA sequences. These additional 109 sequences are
in reality shorter but identical sequences which in most cases varies
only by the presence of leading and trailing gaps. By setting both
<code>clean_dna</code> and <code>collapseSubstrings</code> to
<code>TRUE</code>, we ensure the removal of these leading and trailing
gaps from sequences before collapsing these sequences into unique
haplotypes.</p>
<p><strong>Warning:</strong> be carefull when using
<code>collapseSubstrings</code> when using non-coding genes
(e.g. 12S/16S rRNAs)</p>
<p>To visualize the relationships between these unique haplotypes and
the samples, we may use the <code>haplotype_tbl</code> function:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/haplotype_tbl.html">haplotype_tbl</a></span><span class="op">(</span><span class="va">geophagus</span>, clean<span class="op">=</span> <span class="cn">TRUE</span>, collapseSubstrings <span class="op">=</span> <span class="cn">TRUE</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 3</span></span></span>
<span><span class="co">#&gt;    labels     n_seqs collapsed                                                  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                                      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504318.1     38 MZ504328.1, MZ504313.1, MZ504311.1, MZ504312.1, MZ504309.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504540.1     20 MZ504505.1, MZ504553.1, MZ504554.1, MZ504552.1, MZ504542.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504420.1     19 MZ504417.1, MZ504437.1, MZ504425.1, MZ504427.1, MZ504422.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504488.1     16 MZ504538.1, KU568830.1, JN026709.1, MZ504522.1, MZ504523.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504484.1     15 MZ504496.1, MZ504487.1, MZ504573.1, MZ504560.1, MZ504497.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504462.1     14 MZ504479.1, MZ504477.1, MZ504481.1, MZ504476.1, MZ504463.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504375.1     13 MZ504372.1, MZ504382.1, MZ504381.1, MZ504379.1, MZ504383.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504535.1      8 MZ504515.1, MZ504525.1, MZ504533.1, MZ504534.1, MZ504536.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504393.1      8 MZ504445.1, MZ504413.1, MZ504407.1, MZ504408.1, MZ504410.1…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504400.1      6 MZ504404.1, MZ504401.1, MZ504402.1, MZ504403.1, MZ504399.1 </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
<p>You may also collapse other types of information by using
<code>collapse_others</code>. We can collapse our metadata to see which
species and sampling localities are associated with each haplotype:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haps_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/haplotype_tbl.html">haplotype_tbl</a></span><span class="op">(</span><span class="va">geophagus</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/collapse_others.html">collapse_others</a></span><span class="op">(</span>data<span class="op">=</span> <span class="va">geophagus_info</span>, </span>
<span>                hap_tbl<span class="op">=</span> <span class="va">haps_df</span>, </span>
<span>                labels<span class="op">=</span> <span class="st">"gbAccession"</span>, </span>
<span>                cols<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"scientificName"</span>, <span class="st">"country"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 5</span></span></span>
<span><span class="co">#&gt;    labels     n_seqs collapsed          collapsed_scientific…¹ collapsed_country</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504318.1     38 MZ504328.1, MZ504… Geophagus_proximus     Brazil: Lago do …</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504540.1     20 MZ504505.1, MZ504… Geophagus_altifrons, … Brazil: Tocantin…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504420.1     19 MZ504417.1, MZ504… Geophagus_sp.1         Brazil: Branco- …</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504488.1     16 MZ504538.1, KU568… Geophagus_altifrons, … Brazil: Água Boa…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504484.1     15 MZ504496.1, MZ504… Geophagus_altifrons    Brazil: São Seba…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504462.1     14 MZ504479.1, MZ504… Geophagus_sp.3         Brazil: Iriri- P…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504375.1     13 MZ504372.1, MZ504… Geophagus_megasema     Brazil: Mamirauá…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504535.1      8 MZ504515.1, MZ504… Geophagus_altifrons    Brazil: Maués-Aç…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504393.1      8 MZ504445.1, MZ504… Geophagus_sp.1         Brazil: Xingu- u…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504400.1      6 MZ504404.1, MZ504… Geophagus_sp.1         Brazil: Demeni- …</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ abbreviated name: ¹​collapsed_scientificName</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="running-species-delimitation-analysis">Running species delimitation analysis<a class="anchor" aria-label="anchor" href="#running-species-delimitation-analysis"></a>
</h3>
<p>After preparing your data (alignments and phylogenetic trees), the
next step it to run species delimitation analyses and combine them into
a single <code>tbl_df</code>. Please check <a href="https://legallab.github.io/delimtools/articles/install.html" class="external-link">Installing
Species Delimitation Software</a> to know which sofware
<code>delimtools</code> support. Check the examples below.</p>
<div class="section level4">
<h4 id="abgd">ABGD<a class="anchor" aria-label="anchor" href="#abgd"></a>
</h4>
<p>We can run ABGD by using the <code>abgd_tbl</code> function. This
function requires a PATH for a FASTA file as input for analysis and a
PATH for the ABGD executable:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dat_all_ali</span> <span class="op">&lt;-</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes/geophagus.fasta"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">abgd_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/abgd_tbl.html">abgd_tbl</a></span><span class="op">(</span>infile<span class="op">=</span> <span class="va">dat_all_ali</span>, </span>
<span>                    exe<span class="op">=</span> <span class="st">"/usr/local/bin/abgd"</span>,</span>
<span>                    model<span class="op">=</span> <span class="fl">3</span>,</span>
<span>                    outfolder<span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes"</span><span class="op">)</span>,</span>
<span>                    haps<span class="op">=</span> <span class="va">haps_df</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> ABGD files are located in directory '/home/pedro/Documentos/Projetos/Packages/delimtools/vignettes'.</span></span></code></pre></div>
<p>Then we can check <code>abgd_df</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">abgd_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels      abgd</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504301.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504318.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504341.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504337.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504342.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504304.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504332.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504343.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504315.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504345.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="asap">ASAP<a class="anchor" aria-label="anchor" href="#asap"></a>
</h4>
<p>We can run ASAP by using the <code>asap_tbl</code> function. This
function requires a PATH for a FASTA file as input for analysis and a
PATH for the ASAP executable:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asap_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/asap_tbl.html">asap_tbl</a></span><span class="op">(</span>infile<span class="op">=</span> <span class="va">dat_all_ali</span>,</span>
<span>                    exe<span class="op">=</span> <span class="st">"/usr/local/bin/asap"</span>,</span>
<span>                    haps<span class="op">=</span> <span class="va">haps_df</span><span class="op">$</span><span class="va">labels</span>,</span>
<span>                    outfolder <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes"</span><span class="op">)</span>,</span>
<span>                    model<span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> ASAP files are located in directory '/home/pedro/Documentos/Projetos/Packages/delimtools/vignettes'.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">asap_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels      asap</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504301.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504318.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504332.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504343.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504304.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504342.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504341.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504315.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504345.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504337.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
<p><strong>Note:</strong> You should always indicate full PATH for
<code>infile</code>, <code>exe</code>, and <code>outfolder</code> in
ASAP for full functioning.</p>
</div>
<div class="section level4">
<h4 id="gmyc">GMYC<a class="anchor" aria-label="anchor" href="#gmyc"></a>
</h4>
<p>We can run GMYC analysis by using the <code>splits</code> function
<code>gmyc</code>. This function requires an ultrametric tree file of
class <code>phylo</code> as input for analysis. We can convert our
<code>geophagus_beast</code> treedata object into a <code>phylo</code>
object by using <code><a href="https://rdrr.io/pkg/ape/man/as.phylo.html" class="external-link">ape::as.phylo</a></code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gmyc_res</span> <span class="op">&lt;-</span> <span class="fu">splits</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/splits/man/gmyc.html" class="external-link">gmyc</a></span><span class="op">(</span><span class="fu">as.phylo</span><span class="op">(</span><span class="va">geophagus_beast</span><span class="op">)</span>, method<span class="op">=</span> <span class="st">"single"</span><span class="op">)</span></span></code></pre></div>
<p>Now, lets check the <code>summary</code>:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">gmyc_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; Result of GMYC species delimitation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  method: single</span></span>
<span><span class="co">#&gt;  likelihood of null model:   995.0483</span></span>
<span><span class="co">#&gt;  maximum likelihood of GMYC model:   1010.027</span></span>
<span><span class="co">#&gt;  likelihood ratio:   29.95778</span></span>
<span><span class="co">#&gt;  result of LR test:  3.124278e-07***</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  number of ML clusters:  19</span></span>
<span><span class="co">#&gt;  confidence interval:    18-22</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  number of ML entities:  21</span></span>
<span><span class="co">#&gt;  confidence interval:    19-25</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  threshold time: -0.0164289</span></span></code></pre></div>
<p>GMYC outputs a list containing several results. We can use
<code>gmyc_tbl</code> to extract the clusters in a <code>tbl_df</code>
format:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gmyc_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gmyc_tbl.html">gmyc_tbl</a></span><span class="op">(</span><span class="va">gmyc_res</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">gmyc_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels      gmyc</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> GU701784.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GU701785.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MH780911.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> OR732927.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> JN988869.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> OR732928.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504387.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504388.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504369.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504390.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="bgmyc">bGMYC<a class="anchor" aria-label="anchor" href="#bgmyc"></a>
</h4>
<p>We can run bGMYC analysis by using the <code>bGMYC</code> function
<code>bgmyc.singlephy</code>. This function requires an ultrametric tree
file of class <code>phylo</code> as input for analysis. We can convert
our <code>geophagus_beast</code> treedata object into a
<code>phylo</code> object by using <code><a href="https://rdrr.io/pkg/ape/man/as.phylo.html" class="external-link">ape::as.phylo</a></code>:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bgmyc_res</span> <span class="op">&lt;-</span> <span class="fu">bGMYC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/bGMYC/man/bgmyc.singlephy.html" class="external-link">bgmyc.singlephy</a></span><span class="op">(</span><span class="fu">as.phylo</span><span class="op">(</span><span class="va">geophagus_beast</span><span class="op">)</span>,</span>
<span>                                    mcmc<span class="op">=</span> <span class="fl">11000</span>,</span>
<span>                                    burnin<span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                                    thinning<span class="op">=</span> <span class="fl">100</span>,</span>
<span>                                    t1<span class="op">=</span> <span class="fl">2</span>,</span>
<span>                                    t2<span class="op">=</span> <span class="fu">Ntip</span><span class="op">(</span><span class="fu">as.phylo</span><span class="op">(</span><span class="va">geophagus_beast</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                    start<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">0.5</span>, <span class="fl">50</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>bGMYC outputs a list containing several results. We can use
<code>bgmyc_tbl</code> to extract the clusters in a <code>tbl_df</code>
format using <code>0.05</code> as posterior probability threshold for
clustering samples into species partitions:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bgmyc_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bgmyc_tbl.html">bgmyc_tbl</a></span><span class="op">(</span><span class="va">bgmyc_res</span>, ppcutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">bgmyc_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels     bgmyc</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> GU701784.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GU701785.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> JN988869.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MH780911.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> OR732927.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> OR732928.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ050845.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ051032.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ051706.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ051794.1     2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="ptp-and-mptp">PTP and mPTP<a class="anchor" aria-label="anchor" href="#ptp-and-mptp"></a>
</h4>
<p>We can run both PTP and mPTP analysis by using the
<code>mptp_tbl</code>. This function requires a PATH for a phylogram as
input for analysis, a PATH for the mptp executable, a string specifying
which method to use (“multi” for mPTP; “single” for PTP), and a minimum
branch length (minbrlen) for computations:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mptp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mptp_tbl.html">mptp_tbl</a></span><span class="op">(</span>infile<span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes/geophagus_raxml.nwk"</span><span class="op">)</span>,</span>
<span>                    exe<span class="op">=</span> <span class="st">"/usr/local/bin/mptp"</span>,</span>
<span>                    method<span class="op">=</span> <span class="st">"multi"</span>,</span>
<span>                    minbrlen<span class="op">=</span> <span class="fl">2e-06</span>,</span>
<span>                    outfolder <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes"</span><span class="op">)</span>,</span>
<span>                    delimname <span class="op">=</span> <span class="st">"mptp"</span><span class="op">)</span></span>
<span><span class="co">#&gt; mptp 0.2.5_linux_x86_64, 16GB RAM, 4 cores</span></span>
<span><span class="co">#&gt; https://github.com/Pas-Kapli/mptp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Parsing tree file...</span></span>
<span><span class="co">#&gt; Loaded rooted tree...</span></span>
<span><span class="co">#&gt; Number of edges greater than minimum branch length: 159 / 272</span></span>
<span><span class="co">#&gt; Score Null Model: 681.702507</span></span>
<span><span class="co">#&gt; Best score for multi coalescent rate: 748.771263</span></span>
<span><span class="co">#&gt; LRT computed p-value: 0.000000</span></span>
<span><span class="co">#&gt; Writing delimitation file /home/pedro/Documentos/Projetos/Packages/delimtools/vignettes/geophagus_raxml.nwk.mptp.multi.txt ...</span></span>
<span><span class="co">#&gt; Number of delimited species: 14</span></span>
<span><span class="co">#&gt; Creating SVG delimitation file /home/pedro/Documentos/Projetos/Packages/delimtools/vignettes/geophagus_raxml.nwk.mptp.multi.svg ...</span></span>
<span><span class="co">#&gt; Done...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> mPTP files are located in '/home/pedro/Documentos/Projetos/Packages/delimtools/vignettes'.</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ptp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mptp_tbl.html">mptp_tbl</a></span><span class="op">(</span>infile<span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes/geophagus_raxml.nwk"</span><span class="op">)</span>,</span>
<span>                   exe<span class="op">=</span> <span class="st">"/usr/local/bin/mptp"</span>,</span>
<span>                   method<span class="op">=</span> <span class="st">"single"</span>,</span>
<span>                   minbrlen<span class="op">=</span> <span class="fl">2e-06</span>,</span>
<span>                   outfolder <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org/reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"vignettes"</span><span class="op">)</span>,</span>
<span>                   delimname <span class="op">=</span> <span class="st">"ptp"</span><span class="op">)</span></span>
<span><span class="co">#&gt; mptp 0.2.5_linux_x86_64, 16GB RAM, 4 cores</span></span>
<span><span class="co">#&gt; https://github.com/Pas-Kapli/mptp</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Parsing tree file...</span></span>
<span><span class="co">#&gt; Loaded rooted tree...</span></span>
<span><span class="co">#&gt; Number of edges greater than minimum branch length: 159 / 272</span></span>
<span><span class="co">#&gt; Score Null Model: 681.702507</span></span>
<span><span class="co">#&gt; Best score for single coalescent rate: 747.727196</span></span>
<span><span class="co">#&gt; LRT computed p-value: 0.000000</span></span>
<span><span class="co">#&gt; Writing delimitation file /home/pedro/Documentos/Projetos/Packages/delimtools/vignettes/geophagus_raxml.nwk.mptp.single.txt ...</span></span>
<span><span class="co">#&gt; Number of delimited species: 17</span></span>
<span><span class="co">#&gt; Creating SVG delimitation file /home/pedro/Documentos/Projetos/Packages/delimtools/vignettes/geophagus_raxml.nwk.mptp.single.svg ...</span></span>
<span><span class="co">#&gt; Done...</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> mPTP files are located in '/home/pedro/Documentos/Projetos/Packages/delimtools/vignettes'.</span></span></code></pre></div>
<p>To estimate a minimum branch length for your dataset, use
<code>minbr_len</code>:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/min_brlen.html">min_brlen</a></span><span class="op">(</span><span class="fu">as.phylo</span><span class="op">(</span><span class="va">geophagus_raxml</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Printing 5 smallest tip-to-tip distances in a tree with 137 tips ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; |dist     |  n|</span></span>
<span><span class="co">#&gt; |:--------|--:|</span></span>
<span><span class="co">#&gt; |0.000002 | 12|</span></span>
<span><span class="co">#&gt; |0.000003 |  2|</span></span>
<span><span class="co">#&gt; |0.000004 |  6|</span></span>
<span><span class="co">#&gt; |0.000005 |  2|</span></span>
<span><span class="co">#&gt; |0.001561 |  2|</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="local-minima">Local Minima<a class="anchor" aria-label="anchor" href="#local-minima"></a>
</h4>
<p>We can run the Local Minima analysis by using the <code>spider</code>
function <code>localMinima</code>. This function requires a genetic
distance matrix as input for analysis:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mat</span> <span class="op">&lt;-</span> <span class="fu">dist.dna</span><span class="op">(</span><span class="va">geophagus</span>, model<span class="op">=</span> <span class="st">"raw"</span>, pairwise.deletion <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lmin</span> <span class="op">&lt;-</span> <span class="fu">spider</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spider/man/localMinima.html" class="external-link">localMinima</a></span><span class="op">(</span><span class="va">mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.006828358 0.018994392 0.040791871 0.055999413 0.067151612 0.082105695</span></span></code></pre></div>
<p>We can plot the <code>lmin</code> object to select the threshold. We
will use the first value of the <code>lmin</code> object:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">lmin</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="va">lmin</span><span class="op">$</span><span class="va">localMinima</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="delimtools_files/figure-html/unnamed-chunk-25-1.png" width="662.4"></p>
<p>Since <code>Local Minima</code> output is a numeric vector, we need
to use <code>locmin_tbl</code> to get our species partitions in a
<code>tbl_df</code> format:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">locmin_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/locmin_tbl.html">locmin_tbl</a></span><span class="op">(</span><span class="va">mat</span>, </span>
<span>                        threshold<span class="op">=</span> <span class="va">lmin</span><span class="op">$</span><span class="va">localMinima</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                        haps<span class="op">=</span> <span class="va">haps_df</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">locmin_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels     locmin</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504301.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504318.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504341.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504337.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504342.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504304.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504332.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504343.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504315.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504345.1      1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
<p>We can also use patristic distances to generate species partitions as
well. Using the <code>geophagus_beast</code> tree, we can apply a
threshold to classify tips of the tree by using its branch lengths:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tr_mat</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/cophenetic.phylo.html" class="external-link">cophenetic.phylo</a></span><span class="op">(</span><span class="fu">as.phylo</span><span class="op">(</span><span class="va">geophagus_beast</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">as.dist</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tr_lmin</span> <span class="op">&lt;-</span> <span class="fu">spider</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/spider/man/localMinima.html" class="external-link">localMinima</a></span><span class="op">(</span><span class="va">tr_mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0.1114476 0.2020889 0.3527801</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tr_lmin</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="va">tr_lmin</span><span class="op">$</span><span class="va">localMinima</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="delimtools_files/figure-html/unnamed-chunk-27-1.png" width="662.4"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">treedist_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/locmin_tbl.html">locmin_tbl</a></span><span class="op">(</span><span class="va">tr_mat</span>,</span>
<span>                          threshold <span class="op">=</span> <span class="va">tr_lmin</span><span class="op">$</span><span class="va">localMinima</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                          delimname <span class="op">=</span> <span class="st">"treedist"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">treedist_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels     treedist</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> GU701784.1        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> GU701785.1        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> JN988869.1        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MH780911.1        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> OR732927.1        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> OR732928.1        1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ050845.1        2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ051032.1        2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ051706.1        2</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ051794.1        2</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="fixed-threshold">Fixed threshold<a class="anchor" aria-label="anchor" href="#fixed-threshold"></a>
</h3>
<p>We can use <code>locmin_tbl</code> to create a species partition
using a fixed threshold. For example, we can use a threshold of 2% to
generate species partitions:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">percent_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/locmin_tbl.html">locmin_tbl</a></span><span class="op">(</span><span class="va">mat</span>, threshold <span class="op">=</span> <span class="fl">0.02</span>, haps<span class="op">=</span> <span class="va">haps_df</span><span class="op">$</span><span class="va">labels</span>, delimname <span class="op">=</span> <span class="st">"percent"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">percent_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels     percent</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504301.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504318.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504341.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504337.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504342.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504304.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504332.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504343.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504315.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504345.1       1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
<div class="section level4">
<h4 id="morphology">Morphology<a class="anchor" aria-label="anchor" href="#morphology"></a>
</h4>
<p>We can also turn species ranks or results from morphological analysis
into a species partition by using <code>morph_tbl</code>. For
<code>geophagus</code> dataset, we will use the scientific names in
<code>geophagus_info</code>:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">morph_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/morph_tbl.html">morph_tbl</a></span><span class="op">(</span>labels<span class="op">=</span> <span class="va">geophagus_info</span><span class="op">$</span><span class="va">gbAccession</span>, </span>
<span>                      sppVector<span class="op">=</span> <span class="va">geophagus_info</span><span class="op">$</span><span class="va">scientificName</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">labels</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">haps_df</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">morph_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 2</span></span></span>
<span><span class="co">#&gt;    labels     morph</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504501.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504504.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504510.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504543.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504545.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504488.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504483.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504558.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504484.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504502.1     1</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="joining-species-delimitation-outputs">Joining Species Delimitation Outputs<a class="anchor" aria-label="anchor" href="#joining-species-delimitation-outputs"></a>
</h3>
<p>Since we have several species partitions, we need to merge them into
a single data frame in order to visualize the results. However, since
each method has its own rankings, we need to recode them prior to any
further analysis. For this task, we can use <code>delim_join</code>.
This function takes a <code>list</code> of species partitions as input
in order to merge and recode all species partitions:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_delims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">abgd_df</span>, <span class="va">asap_df</span>, <span class="va">bgmyc_df</span>, <span class="va">gmyc_df</span>, </span>
<span>                   <span class="va">locmin_df</span>, <span class="va">morph_df</span>, <span class="va">mptp_df</span>, <span class="va">ptp_df</span>, </span>
<span>                   <span class="va">percent_df</span>, <span class="va">treedist_df</span><span class="op">)</span></span>
<span></span>
<span><span class="va">all_delims_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/delim_join.html">delim_join</a></span><span class="op">(</span><span class="va">all_delims</span><span class="op">)</span></span></code></pre></div>
<p>We can summarise the <code>all_delims_df</code> object by using the
<code>report_delim</code> function:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/report_delim.html">report_delim</a></span><span class="op">(</span><span class="va">all_delims_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Joined delimitations have a total of <span style="font-weight: bold;">45</span> unique species partitions.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Check below the number of species partitions per method:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; |method   | partitions|</span></span>
<span><span class="co">#&gt; |:--------|----------:|</span></span>
<span><span class="co">#&gt; |gmyc     |         21|</span></span>
<span><span class="co">#&gt; |locmin   |         21|</span></span>
<span><span class="co">#&gt; |abgd     |         19|</span></span>
<span><span class="co">#&gt; |bgmyc    |         19|</span></span>
<span><span class="co">#&gt; |ptp      |         17|</span></span>
<span><span class="co">#&gt; |morph    |         16|</span></span>
<span><span class="co">#&gt; |asap     |         14|</span></span>
<span><span class="co">#&gt; |mptp     |         14|</span></span>
<span><span class="co">#&gt; |percent  |         12|</span></span>
<span><span class="co">#&gt; |treedist |          7|</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 11</span></span></span>
<span><span class="co">#&gt;    labels     abgd  asap  bgmyc gmyc  locmin morph mptp  percent ptp   treedist</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504301.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504318.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504341.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504337.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504342.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504304.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504332.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504343.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504315.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504345.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1   sp44    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
<p>We can also check the match ratio statistics of Ahrens et al (2014)
using <code>match_ratio</code>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/match_ratio.html">match_ratio</a></span><span class="op">(</span><span class="va">all_delims_df</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">match_ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span>n<span class="op">=</span><span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 45 × 5</span></span></span>
<span><span class="co">#&gt;    pairs            delim_1 delim_2 n_match match_ratio</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>              <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> abgd-locmin           19      21      17        0.85</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> bgmyc-gmyc            19      21      17        0.85</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> asap-percent          14      12      10        0.77</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> abgd-bgmyc            19      19      14        0.74</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> abgd-ptp              19      17      13        0.72</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> bgmyc-ptp             19      17      13        0.72</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> asap-ptp              14      17      11        0.71</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> gmyc-locmin           21      21      15        0.71</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> bgmyc-locmin          19      21      14        0.7 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> gmyc-ptp              21      17      13        0.68</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">11</span> asap-morph            14      16      10        0.67</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">12</span> mptp-ptp              14      17      10        0.65</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">13</span> asap-mptp             14      14       9        0.64</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">14</span> bgmyc-morph           19      16      11        0.63</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">15</span> locmin-ptp            21      17      12        0.63</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">16</span> percent-ptp           12      17       9        0.62</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">17</span> abgd-asap             19      14      10        0.61</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">18</span> asap-bgmyc            14      19      10        0.61</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">19</span> abgd-gmyc             19      21      12        0.6 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">20</span> abgd-percent          19      12       9        0.58</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">21</span> bgmyc-percent         19      12       9        0.58</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">22</span> morph-percent         16      12       8        0.57</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">23</span> abgd-morph            19      16       9        0.51</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">24</span> asap-gmyc             14      21       9        0.51</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">25</span> gmyc-morph            21      16       9        0.49</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">26</span> abgd-mptp             19      14       8        0.48</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">27</span> bgmyc-mptp            19      14       8        0.48</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">28</span> gmyc-percent          21      12       8        0.48</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">29</span> morph-ptp             16      17       8        0.48</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">30</span> morph-mptp            16      14       7        0.47</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">31</span> asap-locmin           14      21       8        0.46</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">32</span> gmyc-mptp             21      14       8        0.46</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">33</span> mptp-percent          14      12       6        0.46</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">34</span> locmin-morph          21      16       8        0.43</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">35</span> locmin-percent        21      12       7        0.42</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">36</span> percent-treedist      12       7       4        0.42</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">37</span> locmin-mptp           21      14       7        0.4 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">38</span> asap-treedist         14       7       4        0.38</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">39</span> mptp-treedist         14       7       4        0.38</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">40</span> ptp-treedist          17       7       4        0.33</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">41</span> abgd-treedist         19       7       4        0.31</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">42</span> bgmyc-treedist        19       7       3        0.23</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">43</span> gmyc-treedist         21       7       3        0.21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">44</span> locmin-treedist       21       7       3        0.21</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">45</span> morph-treedist        16       7       2        0.17</span></span></code></pre></div>
<p>By checking the results, all comparisons including treedist have a
low <code>match_ratio</code>. Since species partitions with very low
<code>match_ratio</code> values can impact consensus calculations, we
will remove treedist before visualizing our results</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_delims_df</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">all_delims_df</span>, <span class="op">-</span><span class="va">treedist</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/report_delim.html">report_delim</a></span><span class="op">(</span><span class="va">all_delims_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Joined delimitations have a total of <span style="font-weight: bold;">43</span> unique species partitions.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Check below the number of species partitions per method:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; |method  | partitions|</span></span>
<span><span class="co">#&gt; |:-------|----------:|</span></span>
<span><span class="co">#&gt; |gmyc    |         21|</span></span>
<span><span class="co">#&gt; |locmin  |         21|</span></span>
<span><span class="co">#&gt; |abgd    |         19|</span></span>
<span><span class="co">#&gt; |bgmyc   |         19|</span></span>
<span><span class="co">#&gt; |ptp     |         17|</span></span>
<span><span class="co">#&gt; |morph   |         16|</span></span>
<span><span class="co">#&gt; |asap    |         14|</span></span>
<span><span class="co">#&gt; |mptp    |         14|</span></span>
<span><span class="co">#&gt; |percent |         12|</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 137 × 10</span></span></span>
<span><span class="co">#&gt;    labels     abgd  asap  bgmyc gmyc  locmin morph mptp  percent ptp  </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> MZ504301.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> MZ504318.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> MZ504341.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> MZ504337.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> MZ504342.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> MZ504304.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> MZ504332.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> MZ504343.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> MZ504315.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> MZ504345.1 sp1   sp20  sp1   sp1   sp1    sp36  sp1   sp40    sp1  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 127 more rows</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualizing-results">Visualizing Results<a class="anchor" aria-label="anchor" href="#visualizing-results"></a>
</h3>
<p>We can visualize the results by using <code>delim_autoplot</code>.
This function will plot a phylogenetic tree alongside with the results
contained in <code>all_delims_df</code>. We will use the
<code>geophagus_beast</code> tree to visualize the results:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># customize tip labels of the tree</span></span>
<span><span class="va">tip_tab</span> <span class="op">&lt;-</span> <span class="va">geophagus_info</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gbAccession</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">geophagus_beast</span><span class="op">@</span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>labs<span class="op">=</span> <span class="fu">glue</span><span class="fu">::</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{gbAccession} | {scientificName}"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gbAccession</span>, <span class="va">labs</span>, <span class="va">scientificName</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a customized color palette</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/delim_brewer.html">delim_brewer</a></span><span class="op">(</span>delim<span class="op">=</span> <span class="va">all_delims_df</span>, package<span class="op">=</span><span class="st">"randomcoloR"</span>, seed<span class="op">=</span><span class="fl">42</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="fu"><a href="../reference/delim_autoplot.html">delim_autoplot</a></span><span class="op">(</span><span class="va">all_delims_df</span>, </span>
<span>               <span class="va">geophagus_beast</span>,</span>
<span>               consensus<span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               n_match <span class="op">=</span> <span class="fl">5</span>,</span>
<span>               tbl_labs<span class="op">=</span> <span class="va">tip_tab</span>,</span>
<span>               col_vec<span class="op">=</span> <span class="va">cols</span>,</span>
<span>               hexpand<span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>               widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="delimtools_files/figure-html/unnamed-chunk-34-1.png" width="1152"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pedro Bittencourt, Rupert Collins, Tomas Hrbek.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.2.</p>
</div>

    </footer>
</div>





  </body>
</html>
